chromium_src/ios/web/webui/url_data_source_ios.mm

```diff
@@ -0,0 +1,154 @@
+// Copyright (c) 2024 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "ios/web/public/webui/url_data_source_ios.h"
+
+#include "base/containers/span.h"
+#include "base/no_destructor.h"
+#include "base/strings/strcat.h"
+#include "base/strings/stringprintf.h"
+#include "base/strings/utf_string_conversions.h"
+#include "ios/components/webui/web_ui_url_constants.h"
+#include "services/network/public/mojom/content_security_policy.mojom.h"
+
+namespace {
+
+// A chrome-untrusted data source's name starts with chrome-untrusted://.
+bool IsChromeUntrustedDataSource(const web::URLDataSourceIOS* source) {
+  static const base::NoDestructor<std::string> kChromeUntrustedSourceNamePrefix(
+      base::StrCat({kChromeUIUntrustedScheme, url::kStandardSchemeSeparator}));
+
+  return base::StartsWith(source->GetSource(),
+                          *kChromeUntrustedSourceNamePrefix,
+                          base::CompareCase::SENSITIVE);
+}
+
+}  // namespace
+
+namespace web {
+bool URLDataSourceIOS::ShouldAddContentSecurityPolicy() const {
```

Comment:
why are we creating our own implementation in the base class instead of `URLDataSourceIOSImpl`?